src/Data/Vector/Mutable/Linear.hs

```diff
@@ -58,93 +61,105 @@ import GHC.Exts hiding (fromList)
 import GHC.Stack
 import Prelude.Linear hiding (length, read)
 import qualified Unsafe.Linear as Unsafe
-import qualified Unsafe.MutableArray as Unsafe
-import qualified  Prelude
+import Data.Array.Mutable.Linear (Array)
+import qualified Data.Array.Mutable.Linear as Array
+import qualified Prelude
 
 -- # Core data types
 -------------------------------------------------------------------------------
 
-type MutArr a = MutableArray# RealWorld a
-
 -- | A dynamic mutable vector.
 data Vector a where
-  -- | Vec (current length, memory allocated) marray
-  Vec :: (Int, Int) -> MutArr a -> Vector a
+  Vec :: Int -- ^ Current length
+      -> Array a -- ^ Underlying array
+      #-> Vector a
 
 -- # API: Construction, Mutation, Queries
 -------------------------------------------------------------------------------
 
--- | Allocate a constant vector of a given non-zero size (and error on a bad
+empty :: (Vector a #-> Unrestricted b) -> Unrestricted b
+empty f = Array.empty (\arr -> f (Vec 0 arr))
+
+-- | Allocate a constant vector of a given non-negative size (and error on a bad
 -- size)
-constant :: HasCallStack =>
-  Int -> a -> (Vector a #-> Unrestricted b) -> Unrestricted b
+constant :: Int -> a -> (Vector a #-> Unrestricted b) -> Unrestricted b
 constant size x f
-  | size <= 0 = error ("Trying to construct a vector of size " ++ show size)
-  | otherwise = f $ Vec (size, size) (Unsafe.newMutArr size x)
+  | size < 0 = error ("Trying to construct a vector of size " ++ show size)
+  | otherwise = fromList (replicate size x) f
 
--- XXX: long line below
--- | Allocator from a non-empty list (and error on empty lists)
+-- | Creates a vector from a list.
 fromList :: HasCallStack => [a] -> (Vector a #-> Unrestricted b) -> Unrestricted b
-fromList [] _ = error "Trying to allocate a vector from an empty list"
-fromList xs@(x:_) (f :: Vector a #-> Unrestricted b) =
-  constant (Prelude.length xs) x buildThenRun
-  where
-    buildThenRun :: Vector a #-> Unrestricted b
-    buildThenRun vec = f $ doWrites (zip xs [0..]) vec
-
-    doWrites :: [(a,Int)] -> Vector a #-> Vector a
-    doWrites [] vec = vec
-    doWrites ((a,ix):ws) vec = doWrites ws (write vec ix a)
+fromList xs f =
+  Array.fromList xs Prelude.$ \arr ->
+    Array.size arr & \(arr', Unrestricted i) ->
+      f (Vec i arr')
 
 -- | Length of the vector
-length :: Vector a #-> (Vector a, Int)
-length = Unsafe.toLinear unsafeLength
-  where
-    unsafeLength :: Vector a -> (Vector a, Int)
-    unsafeLength v@(Vec (len, _) _) = (v, len)
+length :: Vector a #-> (Vector a, Unrestricted Int)
```

Comment:
I replaced this function to return an `Unrestricted Int` instead of `Int`, to be more consistent with other functions like `read`.